designs/static-capacity.md

```diff
@@ -0,0 +1,141 @@
+# Karpenter - Static Capacity
+
+## Background
+Karpenter currently operates as a dynamic cluster autoscaler, automatically adjusting node counts based on pending pod demand. However, several important use cases require maintaining a fixed set of nodes:
+
+1. Performance-critical applications where just-in-time provisioning latency is unacceptable
+2. Workloads that need guaranteed capacity regardless of pod demand  
+3. Scenarios where fixed node counts are preferred
+
+Currently, users attempt to achieve static capacity through workarounds:
+- Running placeholder pods to maintain minimum capacity
+- Using separate node management tools alongside Karpenter (MNG/Fargate/Self-managed)
+- Configuring complex provisioner requirements to approximate static behavior
+
+## Proposal
+Extend the existing NodePool resource to support static provisioning capabilities by adding new fields:
+
+```yaml
+spec:
+  # Makes NodePool static when specified
+  replicas: 5  # Number of nodes to maintain
+```
+
+Key aspects:
+1. Static NodePools maintain fixed node count regardless of pod demand
+2. Will not be considered as consolidation candidate (at least in v1alpha1)
+3. Nodes are spread across AZs by default
+4. Inherit existing Karpenter features like drift detection and interruption handling
+5. Use existing NodeClaim infrastructure with owner references
+  
+
+### Modeling & Validation
+
+When `replicas` is specified, the `NodePool` enters a static provisioning mode where certain disruption-related fields become irrelevant or misleading. Specifically:
+
+- `limits` and `weight` (used to influence consolidation and node selection) are not meaningful for static capacity.
+- `consolidationPolicy` and `consolidateAfter` (which control Karpenter’s consolidation logic) must not be used when the node count is fixed via `replicas`.
+
+However, **Kubernetes' schema-level validation (OpenAPI-based) lacks the ability to express conditional logic between fields**. This introduces limitations:
+
+- Non-pointer types defaulting at parse time
+- Default values being indistinguishable from explicitly set values in admission webhooks
+
+### Disruption
+
+Karpenter already has Disruption and Interruption Management for nodes we can use the same mechanism. 
+Since Static Nodepool and NodeClaim is a variant of existing Nodepool/NodeClaim it will inherit Karpenter dynamic provider integration. Drift Detection can also be inherited to trigger replacement of drifted nodes. During this Karpenter will respect the Disruption Budget.
```

Comment:
What do you mean by dynamic provider integration?

Reply:
Supposed to be dynamic provisioning integration, fixed